src/util.rs

```diff
@@ -175,6 +175,16 @@ pub(crate) fn normalize_trailing_zeros(bits: &mut Vec<bool>, bit_len: usize) {
     }
 }
 
+pub(crate) fn concatenate_with_padding<F: PrimeField>(vs: &[Vec<F>], pad_size: usize) -> Vec<F> {
+    let mut vs = vs.to_vec();
+    vs.iter_mut()
+        .flat_map(|v_i| {
+            v_i.resize(pad_size, F::ZERO);
+            v_i.drain(..)
```

Comment:
Why is it so hard? What are you trying to accomplish?

Reply:
I mean, there are a lot of downsides to using such a design, but I want to understand what effects you want to achieve in order to get the review as correct as possible

Reply:
The simplest option with one memory allocation and no cloning at all is suggested, but maybe I missed something
```rust
pub(crate) fn concatenate_with_padding<F: PrimeField>(vs: &[Vec<F>], pad_size: usize) -> Vec<F> {
    vs.iter()
        .fold(Vec::with_capacity(vs.len() * pad_size), |mut result, v| {
            result.extend_from_slice(v);
            result.extend(iter::repeat(F::ZERO).take(pad_size.saturating_sub(v.len())));
            result
        })
}
```

Reply:
No, there is no other concern. This is adapted from my previous code based on suggestion from chatgpt. The only purpose is to avoid extra clone.
What is the downside of using drain?

Reply:
There was a lot of implicit clones there, as well as a lot of unnecessary allocations when expanding vectors.

Reply:
@chaosma Do we need an error, if `vs` element is larger than padding size?